staging/src/k8s.io/client-go/datapartition/apiserverconfigmanager.go

```diff
@@ -0,0 +1,285 @@
+	kubeClient            clientset.Interface
+	recorder              record.EventRecorder
+
+	isApiServerConfigInitialized bool
+	APIServerMap                 map[string]v1.EndpointSubset
+	rev                          int
+
+	updateChGrp *bcast.Group
+
+	mux             sync.RWMutex
+	firstUpdateTime time.Time
+}
+
+func GetAPIServerConfigManager() *APIServerConfigManager {
+	return instance
+}
+
+func checkInstanceExistence() {
+	if instance != nil {
+		klog.Fatalf("Unexpected reference to api server config manager - initialized")
+	}
+}
+
+func NewAPIServerConfigManager(kubeClient clientset.Interface) *APIServerConfigManager {
+	manager := &APIServerConfigManager{
+		kubeClient:                   kubeClient,
+		isApiServerConfigInitialized: false,
+		APIServerMap:                 make(map[string]v1.EndpointSubset),
+	}
+
+	err := manager.syncApiServerConfig()
+	if err != nil {
+		klog.Fatalf("Unable to get api server list from registry. Error %v", err)
+	}
+
+	instance = manager
+	return instance
+}
+
+func NewAPIServerConfigManagerWithInformer(epInformer coreinformers.EndpointsInformer, kubeClient clientset.Interface) *APIServerConfigManager {
+	checkInstanceExistence()
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().EventsWithMultiTenancy("", "")})
+
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("api_server_config_manager", kubeClient.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	manager := &APIServerConfigManager{
+		kubeClient:                   kubeClient,
+		isApiServerConfigInitialized: false,
+		APIServerMap:                 make(map[string]v1.EndpointSubset),
+		updateChGrp:                  apiserverupdate.GetAPIServerConfigUpdateChGrp(),
+	}
+
+	// Endpoint for api server should already existed. No need to listen for add event
+	epInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: manager.updateApiServer,
+		DeleteFunc: manager.deleteApiServer,
+	})
+
+	manager.apiserverLister = epInformer.Lister()
+	manager.apiserverListerSynced = epInformer.Informer().HasSynced
+	err := manager.syncApiServerConfig()
+	if err != nil {
+		klog.Fatalf("Unable to get api server list from registry. Error %v", err)
+	}
+
+	instance = manager
+	return instance
+}
+
+func (a *APIServerConfigManager) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting api server config manager")
+	defer klog.Info("Shutting down api server config manager")
+
+	if !cache.WaitForCacheSync(stopCh, a.apiserverListerSynced) {
+		klog.Info("Api service end points NOT synced %v.")
+		return
+	}
+
+	klog.Infof("Caches are synced for api server end points. [%+v]", a.APIServerMap)
+	<-stopCh
+}
+
+// TODO - check updating from others
+func (a *APIServerConfigManager) GetAPIServerConfig() map[string]v1.EndpointSubset {
```

Comment:
Not sure if we need to add a READ lock here or just a WRITE lock

Reply:
https://medium.com/golangspec/sync-rwmutex-ca6c6c3208a0